app/Community/Actions/BuildActivePlayersAction.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Community\Actions;
+
+use App\Community\Data\ActivePlayerData;
+use App\Data\PaginatedData;
+use App\Models\Game;
+use App\Models\User;
+use Illuminate\Pagination\LengthAwarePaginator;
+use Illuminate\Support\Collection;
+use Illuminate\Support\Str;
+
+class BuildActivePlayersAction
+{
+    public function execute(
+        int $perPage = 300,
+        int $page = 1,
+        ?array $gameIds = null,
+        ?string $search = null,
+    ): PaginatedData {
+        $activePlayersList = (new LoadThinActivePlayersListAction())->execute();
+        $unfilteredTotal = count($activePlayersList);
+
+        // Filter the results.
+        $activePlayersList = $this->filterByGameIds($activePlayersList, $gameIds);
+        $activePlayersList = $this->filterBySearch($activePlayersList, $search);
+
+        // Paginate the results.
+        $offset = ($page - 1) * $perPage;
+        $paginatedPlayers = array_slice($activePlayersList, $offset, $perPage);
+
+        $hydratedPlayers = $this->hydrateResults($paginatedPlayers);
+
+        $paginator = new LengthAwarePaginator(
+            items: $hydratedPlayers,
+            total: count($activePlayersList),
+            perPage: $perPage,
+            currentPage: $page,
+        );
+
+        return PaginatedData::fromLengthAwarePaginator($paginator, unfilteredTotal: $unfilteredTotal);
+    }
+
+    /**
+     * @param array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }> $players
+     *
+     * @return array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }>
+     */
+    private function filterByGameIds(array $players, ?array $gameIds = null): array
+    {
+        if (!$gameIds) {
+            return $players;
+        }
+
+        return array_filter(
+            $players,
+            fn ($player) => in_array($player['game_id'], $gameIds, strict: true)
+        );
+    }
+
+    /**
+     * @param array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }> $players
+     *
+     * @return array<array{
+     *   user_id: int,
+     *   game_id: int,
+     *   username: string,
+     *   display_name: string,
+     *   rich_presence: string,
+     *   game_title: string,
+     * }>
+     */
+    private function filterBySearch(array $players, ?string $search): array
+    {
+        if (!$search) {
+            return $players;
+        }
+
+        $search = Str::lower(trim($search));
+
+        // We should also support logical OR searches, for example:
```

Comment:
"should" sounds like a missing feature, but it looks like you're at least trying to support it.

Reply:
Yep, there was a TODO in front of this comment at one point. Updated in latest to remove the word "should".